File: examples/interaction/typescript-bridge/bridge.mjs
#!/usr/bin/env node
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";

class JsonRpcClient {
  constructor(command, args = []) {
    this.nextId = 1;
    this.pending = new Map();
    this.child = spawn(command, args, {
      stdio: ["pipe", "pipe", "inherit"],
    });
    const lines = createInterface({ input: this.child.stdout });
    lines.on("line", (line) => this.#handleLine(line));
  }

  request(method, params = {}) {
    const id = this.nextId++;
    const payload = { jsonrpc: "2.0", id, method, params };
    const promise = new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
    });
    this.child.stdin.write(`${JSON.stringify(payload)}\n`);
    return promise;
  }

  onNotification(message) {
    if (message.method === "display/event") {
      const event = message.params.event;
      if (event.type === "assistant_message_delta") {
        process.stdout.write(event.text);
      } else if (event.type === "assistant_message_final") {
        process.stdout.write("\n");
      } else if (event.type === "approval_requested") {
        console.error(`approval requested: ${event.approval.approvalId}`);
      }
    }
  }

  #handleLine(line) {
    const message = JSON.parse(line);
    if (message.method) {
      this.onNotification(message);
      return;
    }
    const pending = this.pending.get(message.id);
    if (!pending) {
      return;
    }
    this.pending.delete(message.id);
    if (message.error) {
      pending.reject(new Error(message.error.message));
    } else {
      pending.resolve(message.result);
    }
  }
}

async function main() {
  const command = process.argv[2] ?? "noloong-agent-interaction-server";
  const client = new JsonRpcClient(command, process.argv.slice(3));
  await client.request("initialize", {
    name: "typescript-example-bridge",
    requestedAuthority: ["agent.run", "approval.resolve"],
    requestedUx: {
      displayEvents: true,
      streamText: true,
      editMessage: true,
      markdown: true,
      maxMessageBytes: 8192,
    },
  });
  await client.request("session/create", {
    sessionId: "example",
    profileId: "default",
  });
  await client.request("display/subscribe", {
    sessionId: "example",
    ux: {
      displayEvents: true,
      streamText: true,
      editMessage: true,
      maxMessageBytes: 8192,
    },
  });
  await client.request("agent/prompt", {
    sessionId: "example",
    input: {
      type: "text",
      text: "Say hello from the TypeScript bridge.",
    },
  });
  await client.request("shutdown", {});
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: examples/interaction/python-bridge/bridge.py
#!/usr/bin/env python3
import json
import subprocess
import sys
import threading


class JsonRpcClient:
    def __init__(self, command: list[str]) -> None:
        self.next_id = 1
        self.lock = threading.Lock()
        self.pending: dict[int, tuple[threading.Event, dict]] = {}
        self.process = subprocess.Popen(
            command,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=None,
            text=True,
            bufsize=1,
        )
        self.reader = threading.Thread(target=self._read_loop, daemon=True)
        self.reader.start()

    def request(
        self,
        method: str,
        params: dict | None = None,
        timeout_s: float = 60.0,
    ) -> dict:
        with self.lock:
            request_id = self.next_id
            self.next_id += 1
        event = threading.Event()
        slot: dict = {}
        with self.lock:
            self.pending[request_id] = (event, slot)
        payload = {
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
            "params": params or {},
        }
        assert self.process.stdin is not None
        self.process.stdin.write(json.dumps(payload) + "\n")
        self.process.stdin.flush()
        if not event.wait(timeout_s):
            with self.lock:
                self.pending.pop(request_id, None)
            raise TimeoutError(f"json-rpc request timed out: {method}")
        if "error" in slot:
            raise RuntimeError(slot["error"]["message"])
        return slot["result"]

    def on_notification(self, message: dict) -> None:
        if message.get("method") != "display/event":
            return
        event = message["params"]["event"]
        if event["type"] == "assistant_message_delta":
            print(event["text"], end="", flush=True)
        elif event["type"] == "assistant_message_final":
            print()
        elif event["type"] == "approval_requested":
            approval_id = event["approval"]["approvalId"]
            print(f"approval requested: {approval_id}", file=sys.stderr)

    def _read_loop(self) -> None:
        assert self.process.stdout is not None
        try:
            for line in self.process.stdout:
                message = json.loads(line)
                if "method" in message:
                    self.on_notification(message)
                    continue
                request_id = message["id"]
                with self.lock:
                    pending = self.pending.pop(request_id, None)
                if pending is None:
                    continue
                event, slot = pending
                if "error" in message:
                    slot["error"] = message["error"]
                else:
                    slot["result"] = message["result"]
                event.set()
        except Exception as error:
            self._fail_pending(str(error))
            return
        self._fail_pending("json-rpc server closed stdout")

    def _fail_pending(self, message: str) -> None:
        with self.lock:
            pending = list(self.pending.values())
            self.pending.clear()
        for event, slot in pending:
            slot["error"] = {"message": message}
            event.set()


def main() -> None:
    command = sys.argv[1:] or ["noloong-agent-interaction-server"]
    client = JsonRpcClient(command)
    client.request(
        "initialize",
        {
            "name": "python-example-bridge",
            "requestedAuthority": ["agent.run", "approval.resolve"],
            "requestedUx": {
                "displayEvents": True,
                "streamText": False,
                "markdown": True,
                "maxMessageBytes": 8192,
            },
        },
    )
    client.request("session/create", {"sessionId": "example", "profileId": "default"})
    client.request(
        "display/subscribe",
        {
            "sessionId": "example",
            "ux": {
                "displayEvents": True,
                "streamText": False,
                "maxMessageBytes": 8192,
            },
        },
    )
    client.request(
        "agent/prompt",
        {
            "sessionId": "example",
            "input": {
                "type": "text",
                "text": "Say hello from the Python bridge.",
            },
        },
    )
    client.request("shutdown", {})


if __name__ == "__main__":
    main()
